feat(client-debugger): Redesign debugger UI

alexvy86 opened the pull request:
## Description

- Re-think the UI for the debugger so we can separate things that are container-specific (container data, audience, state transitions) from things that are container-agnostic (e.g. telemetry).
- Expose the whole debugger UI as a single top-level React component so we can reuse it across scenarios (render it inline with the application, or on its own in the browser's DevTools panel).
- Update the rendered-inline scenario of the debugger UI to use the same message-passing approach that we use for the browser extension, to communicate with the core debugger functionality.

![New debugger UI](https://user-images.githubusercontent.com/716334/226050425-a41cc09a-74a7-4558-af48-12de4e9cbd89.gif)

## Reviewer guidance

- All views and (almost all) react code in client-debugger-chrome-extension package is gone now, since we reuse the top-level component exposed from client-debugger-view. Decent number of file deletions and lines of code removed there.
- Removed the test-related npm scripts in client-debugger-chrome-extension for now since tests on the remaining code there seem non-trivial, and having no tests to execute makes jest complain and fail the build.
- A lot of surface area removed from client-debugger-view's API, now that it only has to expose a single top-level component.
- Changes in the views in client-debugger-view should mostly be about:
  - Merging the `_<name>` views into the corresponding `<name>` views. Their purpose was being shareable with the chrome extension but now we share the whole debugger UI as a single component.
  - Using the message-passing approach to communication, taken pretty much verbatim from the corresponding views in client-debugger-chrome-extension (which as mentioned above, are now gone).
- A few new React components (landing page, menu).
- New message pair to request/reply with the list of containers, so we menu can display them in the menu.

alexvy86 committed:
Logging cleanup

alexvy86 committed:
Initial steps towards redesign

alexvy86 committed:
WIP

alexvy86 committed:
wip

alexvy86 committed:
Cleanup

alexvy86 committed:
Missing file header

rohandubal commented:
The structure of left pane and details looks great. In terms of UI library/framework being used, do you have any insight on what Chrome/ Edge use for their debug windows? Is there any existing library that we can use to adopt color scheme/design language?

alexvy86 commented:
> The structure of left pane and details looks great. In terms of UI library/framework being used, do you have any insight on what Chrome/ Edge use for their debug windows? Is there any existing library that we can use to adopt color scheme/design language?

I don't know if there's any standard for that; agree that it would be great to find it if it exists and leverage it.

alexvy86 committed:
WIP - moving to message passing for inline debugger views. Receiving events as arrays for some reason.

alexvy86 committed:
Fix messages being sent as arrays

alexvy86 committed:
Use messageRelay in Telemetry view

alexvy86 committed:
Fix container view not updating when changing containers

alexvy86 committed:
Import/export fixes

alexvy86 committed:
Formatting

Josmithr commented on packages/tools/client-debugger/client-debugger/src/messaging/Messages.ts:
```diff
@@ -8,10 +8,30 @@
  * @public
  */
-export interface IDebuggerMessage<TData = unknown> {
+export interface IBaseDebuggerMessage<TData = unknown> {
```
I wonder if a better naming distinction might be to name this "IDebuggerMessage", and the current `IDebuggerMessage` something like "ISourcedDebuggerMessage" - since the `source` label is the semantic distinction between the two.

alexvy86 replied:
I did exactly that first (with exactly that name😆). I think I then switch tracks so I didn't have to update existing references to `IDebuggerMessage`. But I agree it makes more sense, I'll update for the final version of the PR.

Josmithr reviewed (commented):
(I've only done a top-level review so far. I'll do a deeper code review shortly.)

If we were planning on keeping the `view` library longer term, I would probably prefer to keep its state management in terms of simple events. But since we aren't, I like how this simplifies things for us.

And the UI changes look great! We might have to toy with things a bit to make them appear nicely in the dev tools panel (where the aspect ratio will be more narrow), but that's likely a concern for a future PR.

I'll start the deeper code review now :)

Josmithr commented on packages/tools/client-debugger/client-debugger-chrome-extension/src/devtools/BackgroundConnection.ts:
```diff
@@ -89,9 +89,11 @@ export class BackgroundConnection
 	 */
-	public postMessage(message: IDebuggerMessage): void {
+	public postMessage(message: IBaseDebuggerMessage): void {
+		const sourcedMessage: IDebuggerMessage = message as IDebuggerMessage;
```
Nit: this could be expressed via the following to be atomic:
```typescript
const sourcedMessage: IDebuggerMessage = {
    ...message,
    source,
};
```

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
 export function FluidClientDebuggers(props: FluidClientDebuggersProps): React.ReactElement {
-	const debuggerRegistry: DebuggerRegistry = getDebuggerRegistry();
+	const [containers, setContainers] = React.useState<ContainerMetadata[]>([]);
+	const [menuSelection, setMenuSelection] = React.useState<string>("");
```
Nit: Can we use `undefined` instead of empty strings / arrays here for the initial state? Especially for the container list, I think empty signifies there are no registered debuggers, while undefined signals we haven't requested the list yet.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
+					<ContainerView containerId={containerId} />
+				);
+			break;
+		// TODO: add the Telemetry view here, without ReactContext
```
Is this comment still relevant?

alexvy86 replied:
I don't think so. 🪓 .

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
 export function FluidClientDebuggers(props: FluidClientDebuggersProps): React.ReactElement {
-	const debuggerRegistry: DebuggerRegistry = getDebuggerRegistry();
+	const [containers, setContainers] = React.useState<ContainerMetadata[]>([]);
+	const [menuSelection, setMenuSelection] = React.useState<string>("");
```
Can we provide a default menu selection? Not sure what we want the initial landing page to be, but it would probably be better to pick a default than require the user select one to start seeing data.

alexvy86 replied:
So, the options available to us are one of the containers or the telemetry view. I don't like the telemetry view as the default... The first container in the list would be ok but we have to wait until we get the list to select it (to render it in the first place). I think a "welcome" screen of sorts might be a better option? We could put some brief descriptions of what users should expect to find under each section in the menu, or how to use the debugger (e.g. "if the container list does not show anything or seems out of date, you can click the refresh button to force a refresh").

Josmithr replied:
That's fair, and I guess it also serves to hide some of the async loading from the user, which is nice.

Josmithr replied:
And actually, it might be nice to have a landing page that points to some of our public documentation and stuff.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
+			innerView = <TelemetryView />;
+			break;
+		case "container":
+			// eslint-disable-next-line no-case-declarations
```
We should probably hoist the responsibility of container selection out of this component and into its own. Single-reponsibility (state-wise) components are the way to go in my opinion. If this component is responsible for managing the outer menu selection (containers vs telemetry vs etc.), then a different component should manage the `containers` view and any state it requires.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
 export function FluidClientDebuggers(props: FluidClientDebuggersProps): React.ReactElement {
-	const debuggerRegistry: DebuggerRegistry = getDebuggerRegistry();
+	const [containers, setContainers] = React.useState<ContainerMetadata[]>([]);
+	const [menuSelection, setMenuSelection] = React.useState<string>("");
```
I wonder if it would be worth using a discriminated union here instead of a string. That way we can store any other state associated with the selection (e.g. containerId for the "containers" tab) alongside it.

Otherwise, we probably need to reset the containerId selection any time we change from "containers" to a different selection, etc..

alexvy86 replied:
True, I like the discriminated union approach, I'll move to that.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Menu.tsx:
```diff
@@ -0,0 +1,73 @@
+ * Props for {@link MenuSection}
+ */
+export interface MenuSectionProps {
+	header: string;
```
Docs?

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Menu.tsx:
```diff
@@ -0,0 +1,73 @@
+ * Props for {@link MenuItem}
+ */
+export interface MenuItemProps {
+	onClick: React.MouseEventHandler<HTMLButtonElement>;
```
Nit: Unless we expect the consumers of `MenuItem` to need the pointer event metadata, we can probably simplify the consumer side of this by simply using `() => void` here instead.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Menu.tsx:
```diff
@@ -0,0 +1,73 @@
+/*!
```
Can we move this module into the `components` directory? Organization in this package has slipped a bit, but I would like to strictly have whatever we deem to be our "root" entry-point components live in the root of "src" if possible.

alexvy86 replied:
No problem, moving it.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/MessageRelayContext.ts:
```diff
@@ -23,16 +22,18 @@ export const MessageRelayContext = React.createContext<IMessageRelay | undefined
 			"MessageRelayContext was not defined. Parent component is responsible for ensuring this has been constructed.",
 		);
 	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Is this needed?

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/RenderClientDebugger.tsx:
```diff
@@ -22,13 +24,28 @@ import { FluidClientDebuggers } from "./Debugger";
 		try {
-			ReactDOM.render(React.createElement(FluidClientDebuggers), debuggerElement, resolve);
+			ReactDOM.render(<RootView />, debuggerElement, () => {
+				targetElement.append(debuggerElement);
```
Is this necessary? ReactDOM.render should already do this.

alexvy86 replied:
React will append the <RootView /> DOM tree to `debuggerElement`, but we still need `debuggerElement` appended to the `targetElement` passed to the function, right? I just brought this line from outside for consistency with how RootView.tsx in the chrome extension does it. We could maybe get rid of the `debuggerElement` altogether and just render <RootView /> straight onto the `targetElement`?

alexvy86 replied:
I ended up creating that element outside the function call and passing that instead of the previous arg... which was `document.body`, so rendering the debugger made the app itself diseappear 😄 .

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerHistoryView.tsx:
```diff
@@ -8,47 +8,70 @@ import React from "react";
+			messageRelay.off("message", messageHandler);
 		};
-	}, [clientDebugger, container, setContainerHistory]);
+	}, [containerId, containerHistory, messageRelay, setContainerHistory]);
```
We don't want to include `containerHistory` here, since this effect directly changes it. Otherwise we're rebind all of the listeners and such each time the value changes.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSummaryView.tsx:
```diff
@@ -142,6 +177,7 @@ export interface _ContainerSummaryViewProps extends ContainerStateMetadata, ICon
 export function _ContainerSummaryView(props: _ContainerSummaryViewProps): React.ReactElement {
```
We should be able to collapse this logic back into `ContainerSummaryView` now, right?

alexvy86 replied:
Yes! Hadn't gotten around to that but the underscore views will all go away, merged into their non-underscore versions.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/DataObjectsView.tsx:
```diff
@@ -2,33 +2,35 @@
 	 */
 	renderOptions: SharedObjectRenderOptions;
+
+	containerData: unknown;
```
`DataObjectsView` will be querying for its data via the MessageRelay, right? Is this needed?

Josmithr replied:
Or is this a temporary thing to preserve the existing behavior?

alexvy86 replied:
The prop will go away, yep. I haven't gotten to this view to make the full update so just needed it to build in the meantime.

Josmithr reviewed (commented):
Completed the in-depth code pass. There are a couple of React.useEffect dependency issues, but otherwise my feedback is mostly nitpicks and pattern / naming suggestions.

Thanks for doing this refactoring!!

Josmithr commented:
Styling nitpick: Can we add some sort of highlighting in the container selection list to indicate which one is currently selected?

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/Menu.tsx:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @internal
+ */
+export function MenuSection(props: React.PropsWithChildren<MenuSectionProps>): React.ReactElement {
```
Do you know if syntax exists for that? I couldn't find it. Another alternative would be to explicitly use `FunctionalComponent<Props>` which apparently adds children automatically.

Josmithr replied:
You should be able to do 
```typescript
export interface FooProps extends React.PropsWithChildren {
    ...
}
```

But I have also seen:

```typescript
export type FooProps = React.PropsWithChildren<{
    ...
}>
```

alexvy86 replied:
Ah, the latter will do it. The first one doesn't seem to work, `PropsWithChildren` needs the generic arg:

<img width="656" alt="image" src="https://user-images.githubusercontent.com/716334/226756973-b463a614-3e0b-4d7f-bdde-f40c89ab9887.png">

Josmithr replied:
Interesting. That argument must have become optional at some point (or maybe it went the other way?). I've used the first pattern in a side-project pretty extensively. But I forget what version of React I'm using.

alexvy86 committed:
PR review comments and cleanup

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/MessageRelayContext.ts:
```diff
@@ -23,16 +22,18 @@ export const MessageRelayContext = React.createContext<IMessageRelay | undefined
  * @throws If {@link MessageRelayContext} has not been set.
  */
 export function useMessageRelay(): IMessageRelay {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Nope, probably a complaint from a TS server that failed to refresh. Same below.

alexvy86 committed:
Remove unnecessary things from export

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:
```diff
@@ -29,16 +32,16 @@ export function TelemetryView(): React.ReactElement {
+			messageRelay.off("message", messageHandler);
 		};
-	}, [telemetryEvents, setTelemetryEvents]);
+	}, [messageRelay, telemetryEvents, setTelemetryEvents]);
```
So we ran into a scenario where we know better than the linter and will have to disable the rule? Living on the edge 😝 . But I do get your point.

Josmithr replied:
Oh, I missed that we were actually observing the state of `telemetryEvents` by concatenating the existing ones with the new ones coming in. I think we actually **do** need to list it here, otherwise we could end up concatenating with the wrong `telemetryEvents` state (I think).

This might be a case where we need to wrap the concat logic in a `useCallback` hook, to ensure we're always getting the most up-to-date version of `telemetryEvents`. I'll admit, I don't have a lot of experience with this particular use case, so this might require a bit of investigation.

alexvy86 replied:
[This](https://react.dev/reference/react/useEffect#updating-state-based-on-previous-state-from-an-effect) was the solution. `useCallback` would have had the same problem, I think. The callback function itself would need to be passed as a dependency to the effect, and we end up with the same cycle but in two steps instead of one.

Josmithr replied:
Ah, yeah that would do it!

alexvy86 committed:
Cleanup good-fences config after removed modules

alexvy86 committed:
Fix export complain

alexvy86 committed:
Bug fix: unnecessary react state dep

alexvy86 committed:
Remove unnecessary extra dom element

alexvy86 committed:
Remove underscore views

alexvy86 committed:
Fix container state message passing

alexvy86 committed:
Formatting

alexvy86 committed:
Create dom node for debugger

alexvy86 committed:
Adjust debugger message interface names

alexvy86 committed:
Use discriminated union for menu selection

alexvy86 committed:
Messaging infrastructure for container data

alexvy86 committed:
Landing View

alexvy86 committed:
Fix react effect dependency

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 committed:
Selected and hover styles for menu

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 committed:
Merge _AudienceView into AudienceView

alexvy86 committed:
Visual changes. Ensure background connection initializes before rendering devtools debugger.

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 committed:
Cleanup

alexvy86 committed:
Fix: use messageRelay to ask for telemetry history

alexvy86 committed:
Clean up messaging. Relays/pipes add the source property on both sides.

alexvy86 commented:
Split some changes into #14809 and #14810 .

alexvy86 committed:
Do not export unnecessary constants

alexvy86 committed:
Improve comment

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 committed:
Remove test scripts for now

alexvy86 committed:
Rename constant

alexvy86 committed:
Fix comment

alexvy86 committed:
Remove unnecessary export

alexvy86 committed:
Pass message relay to debugger-rendering function

alexvy86 committed:
Reuse renderClientDebuggerView in the extension

alexvy86 committed:
Cleanup comment

alexvy86 committed:
Cleanup unused code

alexvy86 committed:
Remove console.log

alexvy86 committed:
Revert space removals

alexvy86 commented:
Ok team, review away and let's merge this!

Josmithr commented on packages/tools/client-debugger/client-debugger-chrome-extension/src/devtools/BackgroundConnection.ts:
```diff
@@ -60,14 +61,20 @@ export class BackgroundConnection
+	/**
+	 * Creates an instance of {@link BackgroundConnection}.
+	 *
+	 * @param messageSource - All messages sent through the returned instance's {@link BackgroundConnection.postMessage}
```
Nit: I think if you make this a comment on the inline property, it will bubble up automatically as a param comment for the constructor.

Josmithr commented on packages/tools/client-debugger/client-debugger-chrome-extension/src/devtools/BackgroundConnection.ts:
```diff
@@ -99,11 +106,16 @@ export class BackgroundConnection
 	/**
-	 * Post a message for the debugger to the Background Script.
+	 * Post a message to the Background Script.
+	 * These messages are mostly for the debugger, but some are for the Background Script itself (for initialization).
```
Nit:
```suggestion
	 * @remarks These messages are mostly for the debugger, but some are for the Background Script itself (for initialization).
```

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -33,145 +45,149 @@ export interface FluidClientDebuggersProps {
 	renderOptions?: RenderOptions;
 }
 
+interface ContainerMenuSelection {
```
Docs for these? 😁

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -33,145 +45,149 @@ export interface FluidClientDebuggersProps {
-						aria-describedby={retryButtonTooltipId}
+		<Stack enableScopedSelectors horizontal styles={stackStyles}>
+			<Stack.Item grow={1} styles={menuStyles}>
+				{/* TODO: button to refresh list of containers */}
```
TODO before merge, or sometime in the future?

alexvy86 replied:
My intention was the short-term future :)

Josmithr replied:
Should we file a task to track this then?

alexvy86 replied:
https://dev.azure.com/fluidframework/internal/_workitems/edit/3854, should be part of grooming tomorrow.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/WindowMessageRelay.ts:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @remarks
+ *
+ * While a debugger view rendered in the same page as the application could technically communicate with the
```
This context is great! Thanks for adding this!

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/WindowMessageRelay.ts:
```diff
@@ -0,0 +1,66 @@
+	extends TypedEventEmitter<IMessageRelayEvents>
+	implements IMessageRelay
+{
+	public constructor(private readonly messageSource: string) {
```
Might be worth adding a brief comment to `messageSource` linking to ISourcedMessage for context.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/WindowMessageRelay.ts:
```diff
@@ -0,0 +1,66 @@
+			...message,
+			source: this.messageSource,
+		};
+		globalThis.postMessage(sourcedMessage, "*"); // TODO: verify target is okay
```
I think the TODO here can go away too.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -2,119 +2,99 @@
+import { Waiting } from "./Waiting";
+
+// TODOs:
+// - Special annotation for the member elected as the summarizer
```
Nit
```suggestion
// - Special annotation for the member elected as the summarizer

```

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerHistoryView.tsx:
```diff
@@ -8,79 +8,68 @@ import React from "react";
-		</Stack>
-	);
-}
+	}, [containerId, messageRelay]);
```
We still need `setContainerHistory` here, don't we?

alexvy86 replied:
Yep! Good catch, thanks. While reading React docs I learned that the `set<>()` functions passed in to the component technically never change (they're the exact same function object) and thus never trigger effects even if they're listed as dependencies, so they're not _technically_ necessary, but we've been using them so I think we should keep them. If we make a conscious decision to stop putting them as deps everywhere, we can review that.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerHistoryView.tsx:
```diff
@@ -8,79 +8,68 @@ import React from "react";
-		</Stack>
-	);
-}
+	}, [containerId, messageRelay]);
```
The effect no longer requests the initial batch of state change data from the debugger; we only get fresh updates. I think we probably want a pattern similar to what we do in the audience view case, where we initialize our state to `undefined`, and send an initial request to the debugger for all the history.

alexvy86 replied:
Yep, I guess that part got lost in the move from the extension views. Re-adding it now.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerView.tsx:
```diff
@@ -109,16 +103,17 @@ export function ClientDebugView(props: ClientDebugViewProps): React.ReactElement
 			}}
-			className={clientDebugViewClassName}
+			className={containerViewClassName}
+			key={containerId}
```
Is this key needed for something specific?

alexvy86 replied:
Doesn't look like it, I don't remember how it got here and I agree it doesn't seem necessary. Removing.

alexvy86 committed:
PR feedback

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/DataObjectsView.tsx:
```diff
@@ -2,39 +2,87 @@
+		return (): void => {
+			messageRelay.off("message", messageHandler);
+		};
+	}, [containerId, messageRelay]);
```
I think we still need `setContainerData` here

alexvy86 replied:
Related to the other comment like this, but yep, adding it.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/DataObjectsView.tsx:
```diff
@@ -2,39 +2,87 @@
+
+		messageRelay.on("message", messageHandler);
+
+		// Request state info for the newly specified containerId
```
Should we call `setContainerData(undefined) here to prevent display of stale data while we load the appropriate container's data from the debugger?

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSummaryView.tsx:
```diff
@@ -26,140 +34,105 @@ initializeFluentUiIcons();
+		// Reset Container State data, to ensure we aren't displaying stats for the wrong container while
+		// we wait for a response from the new debugger.
+		// eslint-disable-next-line unicorn/no-useless-undefined
+		setContainerState(undefined);
```
This is a good call. I think we should probably be doing that in our other `containerId`-dependent components as well for this same reason. Especially if the initial data fetch fails, we won't just continue displaying the wrong data to the user.

alexvy86 replied:
True, I'll copy it to the rest of the views.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:
```diff
@@ -5,21 +5,43 @@
 export function TelemetryView(): React.ReactElement {
+	const messageRelay = useMessageRelay();
+
 	const [telemetryEvents, setTelemetryEvents] = React.useState<ITelemetryBaseEvent[]>([]);
```
It might be worth initializing this to `undefined`, and displaying a spinner until we get a response from the debugger. That way we can differentiate the "I haven't received my data yet" from "There have been no telemetry logs yet" cases.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:
```diff
@@ -5,21 +5,43 @@
+	}
+}
+
+function replacer(key, value): unknown {
```
Is this used?

alexvy86 replied:
Added docs and renamed.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:
```diff
@@ -5,21 +5,43 @@
+function mapEventCategoryToBackgroundColor(eventCategory: string): string {
+	switch (eventCategory) {
+		case "generic":
+			return "#b8ebf2";
```
Probably not for this PR, but I wonder if we should be mapping to FluentUI theme values here, rather than explicit colors.

alexvy86 replied:
We totally should. We'll want to make that part of the final visual homologation (@rohandubal, @WayneFerrao).

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/Menu.tsx:
```diff
@@ -0,0 +1,85 @@
+	},
+};
+
+function menuSectionItemStyles(isActive: boolean): IStackStyles {
```
Nit (feel free to disregard): how do you feel about calling this `getMenuSectionItemStyles`, since it's a function rather than a const?

alexvy86 replied:
I gave that exact advice before, so it'd be weird if I didn't take it :D. Changing.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/LandingView.tsx:
```diff
@@ -0,0 +1,29 @@
+				the Fluid Framework. This includes messages coming from every container, and
+				messages coming from sources outside the context of any given container.
+			</p>
+			{/* <p>For more information, go to [TBD: docs page].</p> */}
```
TODO before merging? Or is this for later?

alexvy86 replied:
Later. I was thinking of public docs for the Debugger, which we don't have yet.

Josmithr reviewed (commented):
I left a few more comments. Mostly nitpicks, but I think there are a few lingering React state issues to resolve before merging.

alexvy86 commented:
> I left a few more comments. Mostly nitpicks, but I think there are a few lingering React state issues to resolve before merging.

@Josmithr addressed, all in a single commit so should be easy to look at just these changes VS the version you had seen.

alexvy86 committed:
Replace renderClientDebuggerView entryPoint with RootView

This helps fix the issue I introduced that broke the resizable in the inline-rendered UI (because I messed up with dom nodes managed with React)

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:
```diff
@@ -28,8 +29,16 @@ function mapEventCategoryToBackgroundColor(eventCategory: string): string {
-function replacer(key, value): unknown {
-	// Filtering out properties
+/**
+ * Function to transform the results when JSON-serializing telemetry events for display.
```
This is great! Thanks!!

Josmithr reviewed (approved):
🚀

alexvy86 committed:
Merge branch 'main' into redesign-debugger-ui

alexvy86 committed:
oops

alexvy86 commented:
Tong's first PR is merged and the second one should be soon; I'll wait for those and integrate those changes, then merge this.
